components/brave_rewards/renderer/creator_detection_agent.cc

```diff
@@ -0,0 +1,265 @@
+    {{"github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"www.github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"gist.github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"reddit.com", IDR_CREATOR_DETECTION_REDDIT_BUNDLE_JS},
+     {"www.reddit.com", IDR_CREATOR_DETECTION_REDDIT_BUNDLE_JS},
+     {"twitch.tv", IDR_CREATOR_DETECTION_TWITCH_BUNDLE_JS},
+     {"www.twitch.tv", IDR_CREATOR_DETECTION_TWITCH_BUNDLE_JS},
+     {"twitter.com", IDR_CREATOR_DETECTION_TWITTER_BUNDLE_JS},
+     {"x.com", IDR_CREATOR_DETECTION_TWITTER_BUNDLE_JS},
+     {"vimeo.com", IDR_CREATOR_DETECTION_VIMEO_BUNDLE_JS},
+     {"www.youtube.com", IDR_CREATOR_DETECTION_YOUTUBE_BUNDLE_JS},
+     {"m.youtube.com", IDR_CREATOR_DETECTION_YOUTUBE_BUNDLE_JS}});
+
+bool ShouldRunInIsolatedWorld(int resource_id) {
+  // By default, scripts are loaded into an isolated JS world. However, the
+  // Twitter script needs to access JS objects that are set directly on DOM
+  // nodes in the main world.
+  if (resource_id == IDR_CREATOR_DETECTION_TWITTER_BUNDLE_JS) {
+    return false;
+  }
+  return true;
+}
+
+v8::Isolate* GetIsolate(content::RenderFrame* render_frame) {
+  DCHECK(render_frame);
+  return render_frame->GetWebFrame()->GetAgentGroupScheduler()->Isolate();
+}
+
+struct DetectionScript {
+  std::string source;
+  bool should_run_in_isolated_world = true;
+};
+
+std::optional<DetectionScript> GetDetectionScript(
+    content::RenderFrame* render_frame) {
+  // Only run scripts for the main frame.
+  if (!render_frame || !render_frame->IsMainFrame()) {
+    return std::nullopt;
+  }
+
+  // Only run scripts if the user has enabled Brave Rewards.
+  if (!render_frame->GetBlinkPreferences().brave_rewards_enabled) {
+    return std::nullopt;
+  }
+
+  auto* web_frame = render_frame->GetWebFrame();
+  DCHECK(web_frame);
+  if (web_frame->IsProvisional()) {
+    return std::nullopt;
+  }
+
+  // Only run scripts for known "media platform" sites, and only over HTTPS.
+  GURL origin = url::Origin(web_frame->GetSecurityOrigin()).GetURL();
+  if (!origin.is_valid() || !origin.SchemeIs(url::kHttpsScheme) ||
+      !IsMediaPlatformURL(origin)) {
+    return std::nullopt;
+  }
+
+  auto iter = kScriptMap.find(origin.host_piece());
+  if (iter == kScriptMap.end()) {
+    return std::nullopt;
+  }
+
+  return DetectionScript{
+      .source = blink::Platform::Current()->GetDataResourceString(iter->second),
+      .should_run_in_isolated_world = ShouldRunInIsolatedWorld(iter->second)};
+}
+
+v8::MaybeLocal<v8::Function> CompileDetectionInitializer(
+    v8::Local<v8::Context> context,
+    std::string_view function_body) {
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks_scope(
+      context, v8::MicrotasksScope::kDoNotRunMicrotasks);
+
+  auto* isolate = context->GetIsolate();
+  v8::ScriptCompiler::Source source(gin::StringToV8(isolate, function_body));
+
+  v8::Local<v8::String> parameters[] = {
```

Comment:
Do you mean using string concatenation to wrap the script in a function header and body? That would probably work, but I was trying to avoid any string concatenation (the script bundle is a largish string). We can do it that way if you prefer though.